icicle-compiler/src/Icicle/Avalanche/Statement/Simp/Mutate.hs

```diff
@@ -1,86 +1,70 @@
 
-import P
+import           P
 
-import           Data.Set (Set)
-import qualified Data.Set as Set
 import           Data.Hashable
+import           Data.Set                             (Set)
+import qualified Data.Set                             as Set
 
 
-mutate
-  :: (Hashable n, Eq n)
-  => Statement a n F.Prim
-  -> Statement a n F.Prim
+-- | Replace immutable array updates with mutable ones.
+--   Assuming the AST is in ANF.
+--
+mutate :: (Hashable n, Eq n, Pretty n)
+       => Statement a n F.Prim
+       -> Statement a n F.Prim
 mutate = mutateStmt Set.empty
 
-
-mutateStmt
-  :: (Hashable n, Eq n)
-  => Set (Name n)
-  -> Statement a n F.Prim -> Statement a n F.Prim
-mutateStmt livevars ss
-  = let go = mutateStmt livevars
-    in case ss of
-         If x s1 s2
-           -> let lvs = livevars <> livevarsStmt s1 <> livevarsStmt s2
-              in If (mutateExp lvs x)
-                    (go s1)
-                    (go s2)
-
-         Let n x s
-           -> let lvs = livevars <> livevarsStmt s
-              in  Let n (mutateExp lvs x)
-                        (go s)
-
-         While t n nt x s
-           -> While t n nt (mutateExp livevars x) (go s)
-
-         ForeachInts t n x y s
-           -> ForeachInts t n (mutateExp livevars x)
-                              (mutateExp livevars y)
-                              (go s)
-
-         ForeachFacts fs vt ft s
-           -> ForeachFacts fs vt ft (go s)
-
-         Block stmts
-           -> let f (st:sts) = mutateStmt (livevars <> Set.unions (fmap livevarsStmt sts)) st
-                              : f sts
-                  f []       = []
-              in  Block $ f stmts
-
-         InitAccumulator acc s
-           -> let lvs  = livevars <> livevarsStmt s
-                  acc' = acc { accInit = mutateExp lvs (accInit acc) }
-              in  InitAccumulator acc' (go s)
-
-         Read local acc t s
-           -> Read local acc t (go s)
-
-         -- If n shadows a binding in x, n is not live. Because after this line
-         -- any reference to n will be the n bound here, not the n bound in x.
-         Write n x
-           -> let lvs = Set.delete n livevars
-              in  Write n (mutateExp lvs x)
-
-         Output n t xs
-           -> Output n t (fmap (first (mutateExp livevars)) xs)
-
-         KeepFactInHistory x
-           -> KeepFactInHistory (mutateExp livevars x)
-
-         LoadResumable{} -> ss
-         SaveResumable{} -> ss
-
-
-mutateExp :: (Hashable n, Eq n) => Set (Name n) -> Exp a n F.Prim -> Exp a n F.Prim
+mutateStmt :: (Hashable n, Eq n, Pretty n)
+           => Set (Name n) -- live vars from other nodes that are dependent this node
+                           -- but not reachable from the AST definition.
+                           -- this is a hack so we don't have to construct a proper CFG
+           -> Statement a n F.Prim
+           -> Statement a n F.Prim
+mutateStmt children ss = case ss of
```

Comment:
no actually... I went over the examples thoroughly, the mutate transform is correct, it's not the problem (C result /= Core result). Still I rewrote it with a proper cfg anyway, we might be able to use that code for other analysis in future.